libtransact/src/database/lmdb.rs

```diff
@@ -239,139 +288,85 @@ impl<'a> Iterator for LmdbDatabaseReaderCursor<'a> {
-                _ => DatabaseError::WriterError(format!("{}", err)),
-            })
+    fn put(&mut self, key: &[u8], value: &[u8]) -> Result<(), DatabaseError> {
+        match self.txn.as_ref() {
+            Some(txn) => txn
+                .access()
+                .put(&self.db.main, key, value, lmdb::put::NOOVERWRITE)
+                .map_err(|err| match err {
+                    lmdb::error::Error::Code(lmdb::error::KEYEXIST) => {
+                        DatabaseError::DuplicateEntry
+                    }
+                    _ => DatabaseError::WriterError(format!("{}", err)),
+                }),
+            None => Err(DatabaseError::WriterError(
+                "No lmdb::WriteTransaction found".to_string(),
+            )),
+        }
     }
 
-    pub fn overwrite(&mut self, key: &[u8], value: &[u8]) -> Result<(), DatabaseError> {
+    fn overwrite(&mut self, key: &[u8], value: &[u8]) -> Result<(), DatabaseError> {
         self.txn
+            .as_ref()
+            .unwrap()
             .access()
             .put(&self.db.main, key, value, lmdb::put::Flags::empty())
             .map_err(|err| DatabaseError::WriterError(format!("{}", err)))
     }
 
-    pub fn delete(&mut self, key: &[u8]) -> Result<(), DatabaseError> {
+    fn delete(&mut self, key: &[u8]) -> Result<(), DatabaseError> {
         self.txn
+            .as_ref()
+            .unwrap()
             .access()
             .del_key(&self.db.main, key)
             .map_err(|err| DatabaseError::WriterError(format!("{}", err)))
     }
 
-    pub fn index_put(
-        &mut self,
-        index: &str,
-        key: &[u8],
-        value: &[u8],
-    ) -> Result<(), DatabaseError> {
+    fn index_put(&mut self, index: &str, key: &[u8], value: &[u8]) -> Result<(), DatabaseError> {
         let index = self
             .db
             .indexes
             .get(index)
             .ok_or_else(|| DatabaseError::WriterError(format!("Not an index: {}", index)))?;
         self.txn
+            .as_ref()
+            .unwrap()
             .access()
             .put(index, key, value, lmdb::put::Flags::empty())
             .map_err(|err| DatabaseError::WriterError(format!("{}", err)))
     }
 
-    pub fn index_delete(&mut self, index: &str, key: &[u8]) -> Result<(), DatabaseError> {
+    fn index_delete(&mut self, index: &str, key: &[u8]) -> Result<(), DatabaseError> {
         let index = self
             .db
             .indexes
             .get(index)
             .ok_or_else(|| DatabaseError::WriterError(format!("Not an index: {}", index)))?;
         self.txn
+            .as_ref()
+            .unwrap()
             .access()
             .del_key(index, key)
             .map_err(|err| DatabaseError::WriterError(format!("{}", err)))
     }
 
-    pub fn commit(self) -> Result<(), DatabaseError> {
-        self.txn
+    fn commit(&mut self) -> Result<(), DatabaseError> {
```

Comment:
I could not have the commit method consume self.  When calling this method in merkle, I'd get a compiler error. 

```
error[E0161]: cannot move a value of type dyn database::lmdb::DatabaseWriter: the size of dyn database::lmdb::DatabaseWriter cannot be statically determined
   --> libtransact/src/state/merkle.rs:604:5
    |
604 |     db_writer.commit()?;
    |     ^^^^^^^^^


```
To wrap txn into an option was the solution I could think of. 
I still have to write checks elsewhere to make sure that we do not try to unwrap txn when it's value is None. 
But I'm open for other suggestions on how to best deal with this